port/win/io_win.cc

```diff
@@ -183,25 +183,17 @@ size_t GetUniqueIdFromFile(HANDLE hFile, char* id, size_t max_size) {
 #if (_WIN32_WINNT == _WIN32_WINNT_VISTA)
   // MINGGW as defined by CMake file.
   // yuslepukhin: I hate the guts of the above macros.
-  // This impl does not guarantee uniqueness everywhere
```

Comment:
@yuslepukhin Do you know about the guarantees for the file ID retrieved by `GetFileInformationByHandleEx()` in the other branch of this conditional compilation? In particular, are IDs both unique and guaranteed not to be reused after a file is deleted? (If not, we should probably expand the scope of this change to that branch as well.)

Reply:
It should work for the other branch. This branch should work for NTFS. No guarantees for FAT anywhere. I have not looked for ReFS.

---
In reply to: [327369485](https://github.com/facebook/rocksdb/pull/5844#discussion_r327369485) [](ancestors = 327369485)

Reply:
IDs can be reused by the file system. "File IDs are not guaranteed to be unique over time, because file systems are free to reuse them. In some cases, the file ID for a file can change over time." (See https://docs.microsoft.com/en-us/windows/win32/api/fileapi/ns-fileapi-by_handle_file_information)

Reply:
Thanks for the replies. If the file ID returned by `GetFileInformationByHandleEx()` can be reused, then deleted files' blocks may be read from block cache when attempting to read blocks from newly created files, leading to corruption.

Reply:
I think @ajkr  was asking about GetFileInformationByHandleEx() not GetFileInformationByHandle(), For Ex(), it should guarantee the uniqueness:
https://docs.microsoft.com/en-us/windows/win32/api/winbase/ns-winbase-file_id_info

Reply:
I think the document is saying the 64bits portion of the 128bits cannot guarantee the uniqueness. 
FILE_ID_INFO from GetFileInformationByHandleEx() should be unique and not reused:
https://docs.microsoft.com/en-us/windows/win32/api/winbase/ns-winbase-file_id_info

Reply:
Or only ReFS supports 128 bits? Needs some tests on NTFS and FAT

Reply:
I did some tests on GetFileInformationByHandle()/GetFileInformationByHandleEx() with NTFS, FAT, FAT32, ReFS formats.

- GetFileInformationByHandle() works on all formats, but uniqueness can be only guaranteed if there is no file deletion.

- GetFileInformationByHandleEx()
ReFS: returns 128 bits - uniqueness guarantee
NTFS: returns 64 bits - somewhat unique, id may be reused after file removed.
FAT/FAT32: returns invalid FileInfo.FileId.Identifier, the lowest byte == 1, all higher bytes are zero.

It looks like we need leverage GetFileInformationByHandle to support FAT/FAT32.

Reply:
Thanks very much for finding out each function's guarantees on each filesystem. It should help us decide how to move forward.